buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/RateLimitInternalExtension.java

```diff
@@ -0,0 +1,98 @@
+package org.elasticsearch.gradle.testclusters;
+
+import org.gradle.api.Project;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * This extensions was meant to be used internally by testclusters
+ *
+ * It holds synchronization primitives needed to implement the rate limiting.
+ * This is tricky because we can't use Gradle workers as there's no way to make sure that tests and their clusters are
+ * allocated atomically, so we could be in a situation where all workers are tests waiting for clusters to start up.
+ *
+ * Also auto configures cleanup of executors to make sure we don't leak threads in the daemon.
+ */
+public class RateLimitInternalExtension {
+
+    private static final int EXECUTOR_SHUTDOWN_TIMEOUT = 1;
+    private static final TimeUnit EXECUTOR_SHUTDOWN_TIMEOUT_UNIT = TimeUnit.MINUTES;
+
+    private static final Logger logger =  Logging.getLogger(RateLimitInternalExtension.class);
+
+    private final Semaphore globalSemaphore;
+    private final ExecutorService executorService;
+    private final TestClusterCleanupOnShutdown cleanupThread;
+
+    public RateLimitInternalExtension() {
+        globalSemaphore = new Semaphore(maxPermits());
+        executorService = Executors.newFixedThreadPool(
+            RateLimitInternalExtension.maxPermits() +
+                1
+        );
+        cleanupThread = new TestClusterCleanupOnShutdown();
+        executorService.submit(cleanupThread);
+    }
+
+    public static Semaphore semaphore(Project project) {
+        return getSelfAsExtension(project).globalSemaphore;
+    }
+
+    public static ExecutorService executorService(Project project) {
+        return getSelfAsExtension(project).executorService;
+    }
+
+    public static TestClusterCleanupOnShutdown cleanupThread(Project project) {
+        return getSelfAsExtension(project).cleanupThread;
+    }
+
+    static int maxPermits() {
+        return Optional.ofNullable(System.getProperty("testclusters.max-nodes"))
+            .map(Integer::valueOf)
+            .orElse(Runtime.getRuntime().availableProcessors() / 2);
+    }
+
+    private static RateLimitInternalExtension getSelfAsExtension(Project project) {
```

Comment:
This is a really odd pattern. I don't quite see the benefit of this. Making these methods `static` but requiring an instance of `Project` doesn't increase their convenience and just masks the fact that we _are_ holding project state here.

All this logic should not live in this extension. It should live in a plugin somewhere. Either in `TestClustersPlugin` or in a separate one.

Reply:
Per my other comment around the semaphore access. Our encapsulation here is completely backwards. We have an extension object that manages it's own lifecycle and construction (which should be done by the plugin) and a plugin which contains a bunch of business logic which should really live in this extension. I think perhaps we might need to come to some kind of consensus on best practices here.

Reply:
The primary role of the extension is to hold global state. 
The reason it manages it's own life-cycle is that this allows for most of this complex synchronization to be in a single class without exposing the details, but you're right that it fails at doing so by exposing the executor and semaphore directly. 
I'll create higher level methods and keep those private.